ext/semian/sliding_window.c

```diff
@@ -0,0 +1,458 @@
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval;
+
+  sem_meta_lock(res->sem_id);
+  {
+    retval = RB_INT2NUM(res->shmem->max_size);
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_max_size_set(VALUE self, VALUE new_size)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval;
+
+  int new_max_size = RB_NUM2INT(new_size);
+  if (new_max_size < 1) {
+    rb_raise(rb_eArgError, "max_size must be positive");
+  }
+
+  sem_meta_lock(res->sem_id);
+  {
+    retval = resize_window(res->shmem, new_max_size);
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_values(VALUE self)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval;
+
+  sem_meta_lock(res->sem_id);
+  {
+    retval = rb_ary_new_capa(res->shmem->length);
+    for (int i = 0; i < res->shmem->length; ++i) {
+      int index = (res->shmem->start + i) % res->shmem->max_size;
+      int value = res->shmem->data[index];
+      rb_ary_store(retval, i, RB_INT2NUM(value));
+    }
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_last(VALUE self)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+  VALUE retval;
+
+  sem_meta_lock(res->sem_id);
+  {
+    int index = (res->shmem->start + res->shmem->length - 1) % res->shmem->max_size;
+    retval = RB_INT2NUM(res->shmem->data[index]);
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return retval;
+}
+
+VALUE
+semian_simple_sliding_window_clear(VALUE self)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+
+  sem_meta_lock(res->sem_id);
+  {
+    dprintf("Clearing sliding window");
+    res->shmem->length = 0;
+    res->shmem->start = 0;
+    res->shmem->end = 0;
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return self;
+}
+
+VALUE
+semian_simple_sliding_window_reject(VALUE self)
+{
+  semian_simple_sliding_window_t *res = get_object(self);
+
+  rb_need_block();
+
+  sem_meta_lock(res->sem_id);
+  {
+    // Store these values because we're going to be modifying the buffer.
+    int start = res->shmem->start;
+    int length = res->shmem->length;
+    dprintf("reject! - start:%d end:%d length:%d max_size:%d", res->shmem->start, res->shmem->end, res->shmem->length, res->shmem->max_size);
+
+    int cleared = 0;
+    for (int i = 0; i < length; ++i) {
+      int index = (start + i) % length;
+      int value = res->shmem->data[index];
+      VALUE y = rb_yield(RB_INT2NUM(value));
+      if (RTEST(y)) {
+        if (cleared++ != i) {
+          sem_meta_unlock(res->sem_id);
+          rb_raise(rb_eArgError, "reject! must delete monotonically");
+        }
+        res->shmem->start = (res->shmem->start + 1) % res->shmem->length;
+        res->shmem->length--;
+      }
+    }
+  }
+  sem_meta_unlock(res->sem_id);
+
+  return self;
+}
+
+VALUE
+semian_simple_sliding_window_push(VALUE self, VALUE value)
```

Comment:
Many other circuits just use a number and not a sliding window. What are the advantages of either approach? One huge advantage would be that you could just use the Integer class instead of a 450 line sliding window implementation.

Reply:
To do circuit breaking without a sliding window, you'd have to use fixed time intervals for error counts, and reset the counter at the end of each fixed-time interval. There are several issues with this, off the top of my head:

  * That's a breaking change, compared to the current implementation.
  * You don't open circuits as fast, if a series of failure straddles a reset boundary. There can be cases when the circuit never opens at all (which you could argue is a benefit).
  * A master/slave election might be necessary to determine which worker resets the counter, which could be more complex than the sliding window implementation.

This approach is admittedly more complex than a single integer counter but I'm not sure that it's worth sacrificing the benefits of a sliding window.

It's hard to reason about the effect of this without millisecond-accurate production data.

Reply:
Just to ensure we're on the same page, the alternative approach I'm considering is that you have two integers: (1) Number of failures, (2) Timestamp of when to reset the number of failures.  This makes the code significantly simpler (in Ruby, it's a few lines, but it's quite a bit more here). 

I changed this a few years ago in https://github.com/Shopify/semian/pull/24. It's actually addressing a separate point, which was false positives. I think the reason why you didn't arrive at that conclusion was that the previous solution would refresh the TTL every time it was modified.

Overall, likely the way we should look at circuits (I believe @hkdsun suggested this? Or maybe @csfrancis? Not my idea) is to talk about an "error percentage", i.e. if 10% of requests were errors in some window of time, open the circuits.

Reply:
@pushrax suggested this in our shared doc as well.

My primary motivation for host-based circuits is to address the variable latency problem, where a single worker doesn't see error rates above a threshold but the superposition of all errors on a host shows an anomaly.

I'm not sure it works without some sort of window, because in the implementation before #24, the previous errors would only be purged when there was a period of time since the last observed error with no new errors occurring. If the rate of new errors is greater than the `error_timeout`, then the pseudo-window would grow without bound and the circuit would eventually open. So if I'm reading that code correctly, the circuit opens quickly when there are a lot of errors, slowly-but-surely when there are some errors, and only remains closed when there are fewer than one error per `error_timeout`.

Using that sort of approach in a host-based circuit implementation likely regresses to the slowly-but-surely case, where the circuit constantly opens at a low frequency. Combined with capacity loss during the `half_open_resource_timeout` period, this would be an overall drop in available capacity.